packages/react-devtools-shared/src/backend/renderer.js

```diff
@@ -522,13 +524,129 @@ export function attach(
     typeof setSuspenseHandler === 'function' &&
     typeof scheduleUpdate === 'function';
 
+  // Mapping of fiber IDs to error/warning messages and counts.
+  const fiberToErrorsMap: Map<number, Map<string, number>> = new Map();
+  const fiberToWarningsMap: Map<number, Map<string, number>> = new Map();
+
+  function recordErrorOrWarningOnFiber(
+    type: 'error' | 'warn',
+    args: $ReadOnlyArray<any>,
+    fiberID: number,
+  ): void {
+    const message = format(...args);
+
+    const fiberMap = type === 'error' ? fiberToErrorsMap : fiberToWarningsMap;
+    const messageMap = fiberMap.get(fiberID);
+    if (messageMap != null) {
+      const count = messageMap.get(message) || 0;
+      messageMap.set(message, count + 1);
+    } else {
+      fiberMap.set(fiberID, new Map([[message, 1]]));
+    }
+  }
+
+  function clearErrorsAndWarnings() {
+    const updatedFiberIDs = new Set(fiberToErrorsMap.keys());
+    // eslint-disable-next-line no-for-of-loops/no-for-of-loops
+    for (const key of fiberToWarningsMap.keys()) {
+      updatedFiberIDs.add(key);
+    }
+
+    fiberToErrorsMap.clear();
+    fiberToWarningsMap.clear();
+
+    emitErrorOrWarningUpdates(updatedFiberIDs);
+  }
```

Comment:
Was this function broken? It never cleared `pendingErrorOrWarnings`. (For that matter, neither do the two functions below it.) I'll try to fix this in my next commit.

Reply:
I'm pretty sure it worked for me. Maybe it broken during work on deduplication?

Reply:
I don't think it was visibly broken, just conceptually.